src/main/java/com/google/cloud/tools/managedcloudsdk/internal/extract/ZipExtractorProvider.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.internal.extract;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipArchiveInputStream;
+import org.apache.commons.compress.utils.IOUtils;
+
+/** {@link ExtractorProvider} implementation for *.zip files. */
+public final class ZipExtractorProvider implements ExtractorProvider {
+
+  /** Only instantiated in {@link ExtractorFactory}. */
+  @VisibleForTesting
+  ZipExtractorProvider() {}
+
+  @Override
+  public void extract(
+      Path archive, Path destination, ExtractorMessageListener extractorMessageListener)
+      throws IOException {
+    try (ZipArchiveInputStream in = new ZipArchiveInputStream(Files.newInputStream(archive))) {
+      ZipArchiveEntry entry;
+      while ((entry = in.getNextZipEntry()) != null) {
+        final Path entryPath = destination.resolve(entry.getName());
+        if (extractorMessageListener != null) {
+          extractorMessageListener.message(entryPath.toString());
+        }
+        if (entry.isDirectory()) {
+          if (!Files.exists(entryPath)) {
+            Files.createDirectories(entryPath);
+          }
+        } else {
+          try (OutputStream out = new BufferedOutputStream(Files.newOutputStream(entryPath))) {
```

Comment:
Don't you need to check for setting the executable bits here?  Otherwise `gcloud` and `dev_appserver.py` won't be write.  Looks like you might be able to use `ZipArchiveEntry#getUnixMode()`?

Reply:
I had the code in there, but ran into some problems and since we only use zip for windows I avoided it. Lemme try again for completeness.

Reply:
found out why for anyone interested, from the source : 
```
    /**
     * Retrieves the external file attributes.
     *
     * <p><b>Note</b>: {@link ZipArchiveInputStream} is unable to fill
     * this field, you must use {@link ZipFile} if you want to read
     * entries using this attribute.</p>
     *
     * @return the external file attributes
     */
    public long getExternalAttributes() {
        return externalAttributes;
    }
```

I have to use `ZipFile` instead of `ZipArchiveInputStream` it appears to have to something to do with being able to seek on the file, but I don't know for sure.

Reply:
Ah, ok.  Definitely worth documenting.